### ISSUE TITLE
Browser crashes on large transaction

### ISSUE THREAD

**Issue by talolard** [dfahlander/Dexie.js#869]
Hi,
## tl;dr
(In Chrome73.0.3683.86 @Linux) 
- Within a transaction we're adding ~2M objects (with bulkAdd) across a few tables. 
- When the transaction is about to close the browser crashes. When we run the same code without wrapping it in a transaction "it works", and we see very heavy disk usage (as expected). - Without the transaction there is no disk usage.
- When we use the transaction code in firefox, we get an exception that  "Unhandled Rejection (InvalidStateError): An attempt was made to use an object that is not, or is no longer, usable"
##Details
We're trying to implement full text search along the lines of [this paper](https://cs.uwaterloo.ca/~jimmylin/publications/Lin_ICTIR2015.pdf). We want to use trigrams, and the way the inverted index is structured is such that the primary key is [trigram,docId]. In turn that means that the number of inserts into the inverted index table is very big. 

## Code 

This is an open source project, so full code 
- [without transaction](https://github.com/lighttag/ylabel)
- [with transaction](https://github.com/LightTag/ylabel/tree/with-transaction-to-demo-bug) 

But the relevant part is the code when using a transaction looks like this
```es6
export const addData = async (data,index=0) => {

    const step =10000
    if (data.length <=0){
//Happens outside of the transaction
        return  updateDFTableAfterInsert();

    }
    var startDate = new Date();

    return   db.transaction('rw',[POSTINGS_SCHEMA,DATA_SCHEMA],tx=>{
        return addDocsToStore(data.slice(0,step))
        .then(()=>{
            var endDate = new Date();
            console.log(`Inserted ${step} docs in ${endDate-startDate} ms`)
            
        })
    
    })
    .then(()=>{
   //Recursively call the addData function
        return  addData(data.slice(step,data.length))
    })
    
    }

```

And the addDocsToStore function looks like this
```es6
export const addDocsToStore = async (docs)=>{
    let t1  = new Date()
        await dataTable.bulkAdd(docs);
    let t2  = new Date()
    const insertDUration = t2-t1;
    console.log(`It took ${insertDUration} to insert ${docs.length} docs`)
    t1  = new Date()
    const postings = docs.flatMap(doc=>{
        const trigramCounts = getTrigramsCount(doc.content)
        const postings = Object.entries(trigramCounts).map(([trigram, freq]) => ({ trigram, docId:doc.id, freq }));
        return postings;
    })
    t2  = new Date()
    console.log(`It tokk ${t2-t1} to make ${postings.length} postings`);
    const startTime = new Date();
    const step =10000;
    let r1;
    for (let i=0;i<=postings.length;i+=step){
        t1  = new Date()
        r1 =  await postingsTable.bulkAdd(postings.slice(i,i+step))
        t2  = new Date()
        console.log(`Inserted ${step} postings in ${t2-t1} ms. ${postings.length-i} remaining`)
    }
    
    const endTime = new Date();
    console.log(`Added ${postings.length}  in ${endTime-startTime} ms for doc ${docs.length}`)
    return r1;

}
```

## Disk usage
When running the code above, during the part that's inserting postings, NMON shows our cpu usgae and disk usgae like this
![postings](https://user-images.githubusercontent.com/5352830/60875332-edebcd00-a239-11e9-8a5f-2a8cf618b36b.gif)

Whereas when running the code without the transaction scope we see this (high disk write):
![postingsNoTransaction](https://user-images.githubusercontent.com/5352830/60875457-33a89580-a23a-11e9-8ca0-741dfcb22e6a.gif)

## Guess
Based on the above, we're guessing that with the transaction the DB is trying to commit a lot of data all at once and is crashing the browser. 

# Thanks
BTW Dexie is amazing.

**Comment by dfahlander** [comment]
There might also be a (non-documented) limit on how many calls to IDBObjectStore.add() that can be done in parallell (as bulkAdd does). So the best solution for you right now would be to keep calling bulkAdd() but do it in chunks. 

I did fix one think in bulkAdd(), bulkPut() and bulkDelete() just now that would allow the garbage collector to release the provided arrays of objects more eagerly if the calling code also stops referring to the provided array after having called bulkAdd(). But I would anyway recommend to call bulkAdd() in chunks, wait for the promise and then bulkAdd() next chunk, allowing the garbage collector to release the previous chunk.

**Comment by dfahlander** [comment]
Still, without my fix, there is nothing that Dexie keeps in memory on the transaction itself, but it might be the native implementation of IndexedDB that holds all changes in memory until transaction commits, but I hope that is not the case.

**Comment by szucsnandor** [comment]
Hello!

Is this still an issue? When i try to bulkPut a large amount of data it crashes the browser. an array of 1000+ objects each with hundred properties, but no nesting.

**Comment by dfahlander** [comment]
@szucsnandor It's not an issue in Dexie what we know, but a limitation of resources. My advice is still to split large amounts of data into chunks before calling bulkPut or bulkAdd. The number of objects in each chunk depends on the size of each object. There is no recommendation other than testing out a decent number.

**Comment by szucsnandor** [comment]
@dfahlander Thank you, I found the problem. It was the dexie observable. I am using angular and i subscribed to dexie observable in ngOnInit lifecycle.  In this subscription I called toArray on the table i was writing into. I guess a bulkPut is just iterating over the objects and writing them one by one in the database. But that means a change event is fired for every object in the array  bulkPut started to write. This means I called toArray for every object. That used up the memory quickly.
Looks like piping the observable with RxJs last() operator solved the problem for now, i am still testing, but i might have to fire an event manually when I know there is a change.

**Comment by dfahlander** [comment]
Good you found it. Yes, dexie-observable uses a lot of resources, especially when using put() or bulkPut(). You'd see less resources taken with bulkAdd(). If you want a better solution for reacting to database changes, you can use the new DBCore API ([db.use()](https://dexie.org/docs/Dexie/Dexie.use())) to get notified on all mutations. This will take almost no extra resources, in contrast to [the updating hook] (https://dexie.org/docs/Table/Table.hook('updating')) that dexie-observable uses, which slows down put and bulkPut calls significantly as it has to read existing data prior to doing the put operations.

**Comment by szucsnandor** [comment]
Thank you very much, I have a look at it.
